File: RaveX-ravex146/src/main/java/ravex/modules/misc/PacketLogger.java
package ravex.modules.misc;

import ravex.modules.Category;
import ravex.modules.Module;
import ravex.parameter.BooleanParameter;
import net.minecraft.client.Minecraft;
import net.minecraft.network.chat.Component;

public class PacketLogger extends Module {
    public static final PacketLogger INSTANCE = new PacketLogger();

    public final BooleanParameter outgoing = new BooleanParameter("Outgoing", true);
    public final BooleanParameter incoming = new BooleanParameter("Incoming", false);
    public final BooleanParameter chat = new BooleanParameter("To Chat", true);

    private PacketLogger() {
        super("PacketLogger", Category.MISC);
        addParameter(outgoing);
        addParameter(incoming);
        addParameter(chat);
    }

    public void logPacket(String direction, Object packet) {
        if (!getEnabled()) return;
        String name = packet.getClass().getSimpleName();
        String message = "§7[§6Packet§7] §d" + direction + " §e" + name;

        Minecraft mc = Minecraft.getInstance();
        if (chat.getValue() && mc.player != null) {
            mc.player.displayClientMessage(Component.literal(message), false);
        } else {
            System.out.println("[PacketLogger] " + direction + " " + packet.getClass().getName());
        }
    }
}

File: src/main/kotlin/volthack/modules/misc/PacketLogger.kt
package volthack.modules.misc

import net.minecraft.network.protocol.common.ServerboundKeepAlivePacket
import net.minecraft.network.protocol.common.ServerboundPongPacket
import net.minecraft.network.protocol.common.ClientboundKeepAlivePacket
import net.minecraft.network.protocol.common.ClientboundPingPacket
import net.minecraft.network.protocol.game.ServerboundMovePlayerPacket
import net.minecraft.network.protocol.game.ClientboundMoveEntityPacket
import volthack.event.PacketReceiveEvent
import volthack.event.PacketSendEvent
import volthack.manager.PacketManager
import volthack.setting.Category
import volthack.setting.Module
import volthack.util.chat.ChatUtils

object PacketLogger : Module("PacketLogger", "Logs sent and received network packets", Category.MISC) {
    private val logSent by boolean("Log Sent", true, "Log outgoing packets")
    private val logReceived by boolean("Log Received", false, "Log incoming packets")
    private val chatLog by boolean("Log to Chat", true, "Print packets to client chat")
    private val consoleLog by boolean("Log to Console", false, "Print packets to system logs")
    private val ignoreMovement by boolean("Ignore Movement", true, "Ignore player movement packets")
    private val ignoreKeepAlive by boolean("Ignore KeepAlive", true, "Ignore keep-alive, ping, and pong packets")

    private val sendListener = { event: PacketSendEvent ->
        if (enabled && logSent) {
            val packet = event.packet
            if (shouldLog(packet)) {
                logPacket("Sent", packet::class.java.simpleName)
            }
        }
    }

    private val receiveListener = { event: PacketReceiveEvent ->
        if (enabled && logReceived) {
            val packet = event.packet
            if (shouldLog(packet)) {
                logPacket("Received", packet::class.java.simpleName)
            }
        }
    }

    init {
        PacketManager.registerSendListener(sendListener)
        PacketManager.registerReceiveListener(receiveListener)
    }

    private fun shouldLog(packet: Any): Boolean {
        if (ignoreMovement && (packet is ServerboundMovePlayerPacket || packet is ClientboundMoveEntityPacket)) {
            return false
        }
        if (ignoreKeepAlive && (
            packet is ServerboundKeepAlivePacket || 
            packet is ServerboundPongPacket ||
            packet is ClientboundKeepAlivePacket ||
            packet is ClientboundPingPacket
        )) {
            return false
        }
        return true
    }

    private fun logPacket(direction: String, name: String) {
        val dirColor = if (direction == "Sent") "§a" else "§c"
        val message = "§7[§6PacketLogger§7] $dirColor$direction: §f$name"
        
        if (chatLog) {
            ChatUtils.addChatMessage(message)
        }
        if (consoleLog) {
            println("[PacketLogger] $direction: $name")
        }
    }
}
